image/docker_schema2.go

```diff
@@ -173,6 +168,50 @@ func (m *manifestSchema2) UpdatedImage(options types.ManifestUpdateOptions) (typ
 	return memoryImageFromManifest(&copy), nil
 }
 
+// Edits the layers in the manifest when called by replacing them with the
+// appropriate infos provided. This is metadata-only -- the actual layers still
+// have to get to the image.
+func (m *manifestSchema2) performEdits(infos []types.BlobInfo) error {
+	infolen := len(infos)
+
+	if infolen < 1 {
+		return errors.New("cannot edit an image down to less than one layer")
+	}
+
+	// regurgitate the image configuration for recalculation of layers in the event the layer list has been edited.
+	imageConfig := &image{}
+	blob, err := m.ConfigBlob()
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(blob, imageConfig); err != nil {
+		return err
+	}
+
+	imageConfig.History = make([]imageHistory, infolen)
+	imageConfig.RootFS.DiffIDs = make([]digest.Digest, infolen)
+	m.LayersDescriptors = make([]descriptor, infolen)
+	for i, info := range infos {
+		imageConfig.History[i] = imageHistory{}
+		imageConfig.RootFS.DiffIDs[i] = info.Digest
```

Comment:
(See also the `UpdatedImageNeedsLayerDiffIDs` hack. That decision mechanism would obviously not work with the “anything can be returned” hook, and it is also too late by that time; so “should we compute DiffIDs” would need a complete rethinking.)